src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,411 @@
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:cache", cacheLocation);
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+  } catch(std::exception &e) {
+    return -1;
+  }
+  
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:weight", boost::lexical_cast<int>(weight));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_min_avg_weight(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "minAvgWeight:cache");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_min_avg_weight(0, dir->cct->_conf->rgw_local_cache_address, y)) { /* Initialize minimum average weight */
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "minAvgWeight:weight");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+CacheBlock LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  if (entries_map.empty())
+    return {};
+
+  auto it = std::min_element(std::begin(entries_map), std::end(entries_map),
+			      [](const auto& l, const auto& r) { return l.second->localWeight < r.second->localWeight; });
+
+  /* Get victim cache block */
+  std::string key = it->second->key;
+  CacheBlock victim;
+
+  victim.cacheObj.bucketName = key.substr(0, key.find('_')); 
+  key.erase(0, key.find('_') + 1);
+  victim.cacheObj.objName = key.substr(0, key.find('_'));
+  victim.blockID = boost::lexical_cast<uint64_t>(key.substr(key.find('_') + 1, key.length()));
+
+  if (dir->get(&victim, y) < 0) {
+    return {};
+  }
+
+  return victim;
+}
+
+void LFUDAPolicy::shutdown() {
+  dir->shutdown();
+  
+  // call cancel() on the connection's executor
+  boost::asio::dispatch(conn->get_executor(), [c = conn] { c->cancel(); });
+}
+
+int LFUDAPolicy::exist_key(std::string key, optional_yield y) {
+  if (entries_map.count(key) != 0) {
+    return true;
+  }
+
+  return false;
+}
+
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  response<std::string> resp;
+  int age = get_age(y);
+
+  if (exist_key(block->cacheObj.objName, y)) { /* Local copy */
+    auto it = entries_map.find(dir->build_index(block));
```

Comment:
in `entries_map`, we are inserting a `key`, which is being passed in by the filter driver which again is the `oid` arrived at by concatenating the bucket name, object name, length and offset of block etc. It is in the directory that we search using `build_index(block)` I belive

Reply:
Since it is causing a lot of confusion, we can have a common naming scheme here - using `build_index` for arriving at the key, for using in both the Cache Driver calls as well as for using in the Block Directory calls. In CacheBlock there already is a member called `size`. In `build_index` the `size` can also be concatenated to arrive at a key. The reason why I want the `size` also concatenated is because it helps in debugging when we list the contents of the data cache - from the name w can see what size the block is, and it will also help in caching blocks of different sizes (with same object name, bucket name and offset) specially in case of ranged requests.